src/fmllauncher/java/net/minecraftforge/fml/loading/progress/StartupMessageManager.java

```diff
@@ -88,22 +88,41 @@ long getTimestamp() {
         }
     }
 
+    private synchronized static void addMessage(MessageType type, String message, int maxSize)
+    {
+        Map<MessageType, List<Message>> messages = StartupMessageManager.messages;
+
+        List<Message> oldMessagesForType = messages.getOrDefault(type, ImmutableList.of());
+        List<Message> newMessagesForType = ImmutableList.<Message>builder()
+                .addAll(maxSize < 0 ? oldMessagesForType : oldMessagesForType.subList(0, Math.min(oldMessagesForType.size(), maxSize)))
+                .add(new Message(message, type))
+                .build();
+
+        ImmutableMap.Builder<MessageType, List<Message>> builder = ImmutableMap.builder();
```

Comment:
I would suggest keeping this an EnumMap, so we keep the enum-optimized behaviour. The map is not visible outside of this class anyway, so we don't have to worry about people misusing it.